### PR TITLE
added exp and nbf properties to successful verification return

### LINKED ISSUE #52: Is it possible to get the exp and nbf from the pass where they can be read?
Similar to #50 is it possible to return the exp and nbf where they can be read?  (Sorry I had a go at coding it based on #50 but got lost pretty quickly).

## Expected Behavior
Return the exp and nbf dates if you can determine them from the QR code so I can put nice messages like (your pass expires in under a month, go to blah to get a renew).

## Actual Behavior
There isn't a exp or nbf value returned.

## Steps to Reproduce the Problem

1. Call verifyPassURI with a valid | bad exp | bad nbf pass
2. There isn't a exp and/or nbf in the result

## Specifications

- Version: Node 1.0.2
- Platform: Windows

### PR DESCRIPTION
**Pull request by J-eld**
Fixes #52 

successful verification now returns exp and nbf values.

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by noway** [comment]
Maybe we want to return the whole cwt claims object? I wonder.

@wseagar thoughts on this? You think it's a good data shape?

**Comment by adamreisnz** [comment]
If I may join the conversation, I would say returning the original claims object would have its benefits. The data structure might not be ideal, but at least this way any data in there would be accessible for further processing, and even if the spec changes or additional data is added later, this package wouldn't have to change and the new data would still be accessible for consumers 👍🏼

File: README.md
Comment (adamreisnz) [inline README.md:64]: I would recommend calling this `expires` since we're using full words for property names

File: README.md
Comment (adamreisnz) [inline README.md:65]: I would recommend calling this `validFrom` since we're using full words for property names

File: src/main.ts
Comment (adamreisnz) [inline src/main.ts:80]: It'd be great if we can add a `claims: cwtClaims` property here with the full claims data

Comment (noway) [reply src/main.ts:80]: In that case we should just returns `cwtClaims` imo, since `exp` and `nbf` are inside.

**Comment by noway** [comment]
We might have to break compatibility and switch to returning `cwtClaims` (which has all the info inside, credentialsSubject, exp, nbf, etc) instead of `credentialsSubject`.

**Comment by adamreisnz** [comment]
I'd be okay with that, since we transform the data that is returned in the result for our own needs anyway, so might as well skip that step in this library and just return the cwtCredentials directly.

**Comment by wseagar** [comment]
In hindsight it was a mistake for us not to include nbf and exp in the response.

I understand in principal about returning the whole cwtCredentials object but also

`nzcpResult.cwtCredentials.vc.credentialsSubject.givenName` 

Is pretty ugly and the object does contain a bunch of a stuff a verifying app doesn't need to know about. Just from a design principal I like to keep the things that people use the most easiest to grab out of an API.

I think as compromise is lets return do the following:

1. Keep @adamreisnz suggestions of renaming expires/validFrom so its more readable.
2. Then lets also expose the raw decoded object in a field named `decoded` or `raw`

This way we're not breaking backwards compatibility, we are exposing the whole decoded object and keeping the API simple to pickup and use.

```
type VerificationResult = { 
  success: true; 
  violates: null; 
  expires: number | null; 
  validFrom: number | null; 
  credentialSubject: CredentialSubject 
  decoded: CwtClaims
}
```

Thoughts @noway @adamreisnz ?

**Comment by noway** [comment]
@wseagar I agree, makes sense. Great solution on keeping backwards compatibility + having the raw object available. The field names look good to me 👍🏻

**Comment by adamreisnz** [comment]
Yep, I agree that would be a great solution 👍🏼

**Comment by noway** [comment]
@J-eld do you wanna change your PR to match the spec Will laid out? Otherwise I can jump in and sort this out on Saturday.

**Comment by J-eld** [comment]
> @J-eld do you wanna change your PR to match the spec Will laid out? Otherwise I can jump in and sort this out on Saturday.

Yup sure thing, I'll update the PR tonight

**Comment by adamreisnz** [comment]
Great looking forward to it. FYI we'll be using this library for our [Hello Club](https://helloclub.com) member self-verification process for verifying Vaccine Passes. We were looking at building and publishing one ourselves, but saw you beat us to it already haha 👍🏼

**Comment by wseagar** [comment]
Thanks @J-eld it looks great. Just one final point, do we think its worthwhile returning the

`expires` and `validFrom` fields as a javascript date object rather than the unix timestamp? Also are these fields every going to actually be null?

**Comment by adamreisnz** [comment]
Happy for either, we convert them to luxon Datetime objects anyway for
validation. But javascriot date objects would be good :+1:

**Comment by J-eld** [comment]
> `expires` and `validFrom` fields as a javascript date object rather than the unix timestamp? Also are these fields every going to actually be null?

Agreed, I think a date object would be better, I'll change that now, and looks like cwtClaims.exp and nbf can be number or undefined so should I change null to undefined?

**Comment by noway** [comment]
@J-eld I've resolved optionality in types in regards to `expires`, `validFrom` etc. I'm just trying to push to this branch, can you please allow edits from maintainers?


Cheers.

**Comment by J-eld** [comment]
Allow edits checkbox is already ticked and I've sent you a collab invite to my fork as well, hopefully works now.

**Comment by noway** [comment]
@J-eld thanks, was my bad with GitHub CLI. 

I've resolved the optionality ambiguity in the result. 

Old:
```typescript
  | { success: true; violates: null; expires: Date | undefined; validFrom: Date | undefined; credentialSubject: CredentialSubject; raw: CWTClaims | Partial<CWTClaims> }
  | { success: false; violates: Violates; credentialSubject: CredentialSubject | null; };
```

New:
```typescript
  | { success: true; violates: null; expires: Date; validFrom: Date; credentialSubject: CredentialSubject; raw: CWTClaims }
  | { success: false; violates: Violates; expires: Date | null; validFrom: Date | null; credentialSubject: CredentialSubject | null; raw: UnvalidatedCWTClaims | null };
```

So valid result will always have `expires`, `validFrom` and `raw`. Invalid result might have `expires`, `validFrom` and `raw`.

I'll merge this tomorrow and create 1.1.0 release unless there are any objects @J-eld @wseagar @adamreisnz .

**Comment by noway** [comment]
I've also added test coverage

**Review by noway — APPROVED** [review]
LGTM, will merge tomorrow if there are no objections

**Comment by adamreisnz** [comment]
This works brilliantly now, thanks for implementing! 💯

**Pull request merged**
